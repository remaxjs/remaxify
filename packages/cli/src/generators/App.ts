import * as t from '@babel/types';
import { addNamed } from '@babel/helper-module-imports';
import generator from '@babel/generator';
import traverse from '@babel/traverse';
import prettier from 'prettier';
import * as sander from 'sander';
import * as path from 'path';
import { options } from '../options';
import { RUNTIME_PACKAGE } from '../constants';
import { File } from '@babel/types';
import { SourceType } from '../sourceType';

const REMAXIFY_APP = 'RemaxifyApp';
const App = SourceType.App;

export default function generate(ast: File | undefined, filename: string): OutputFile {
  if (!ast) {
    throw new Error('App 文件无法分析 ast');
  }

  filename = path.join(options.rootDir, filename);

  traverse(ast, {
    CallExpression(nodePath) {
      const { node } = nodePath;

      if (!t.isIdentifier(node.callee)) {
        return;
      }

      if (node.callee.name !== App) {
        return;
      }

      const binding = nodePath.scope.getBinding(App);

      if (binding) {
        return;
      }

      const appId = addNamed(nodePath, App, RUNTIME_PACKAGE, { nameHint: REMAXIFY_APP });
      nodePath.replaceWith(t.callExpression(appId, node.arguments));
      nodePath.stop();
    },
  });

  const code = generator(ast).code;
  const content = prettier.format(code, {
    parser: 'babel',
  });

  if (!options.dry) {
    sander.writeFile(options.output, filename, content);
  }

  return {
    filename,
    content,
  };
}
