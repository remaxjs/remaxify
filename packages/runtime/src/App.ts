import * as React from 'react';

export default function App(appConfig: { [key: string]: any }): React.ComponentType {
  return class App extends React.Component {
    constructor(props: any) {
      super(props);

      Object.keys(appConfig).forEach(key => {
        if (typeof appConfig[key] === 'function') {
          (this as any)[key] = appConfig[key].bind(this);
          return;
        }

        (this as any)[key] = appConfig[key];
      });
    }

    public render() {
      return this.props.children;
    }
  };
}
