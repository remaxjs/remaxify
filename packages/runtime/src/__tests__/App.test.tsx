import * as React from 'react';
import TestRenderer from 'react-test-renderer';
import App from '../App';

describe('App', () => {
  it('works', () => {
    const mockShow = jest.fn();
    const config: any = {
      onShow(...params: any) {
        mockShow(...params);
      },
      globalData: {
        data: 'data',
      },
    };

    const AppComponent = App(config);

    const testRenderer = TestRenderer.create(
      <AppComponent>
        <view>app</view>
      </AppComponent>
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();

    const options = { optionA: 'A' };
    testRenderer.root.instance.onShow(options);
    expect(mockShow).toBeCalledWith(options);
    expect(testRenderer.root.instance.globalData.data).toBe('data');
  });
});
