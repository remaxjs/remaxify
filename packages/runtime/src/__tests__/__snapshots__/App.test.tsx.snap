// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App works 1`] = `
<view>
  app
</view>
`;
